crates/astria-sequencer/src/mempool.rs

```diff
@@ -39,167 +62,236 @@ impl PartialOrd for TransactionPriority {
-    pub(crate) async fn insert_all(
-        &mut self,
-        txs: Vec<(SignedTransaction, TransactionPriority)>,
-    ) -> anyhow::Result<()> {
-        for (tx, priority) in txs {
-            self.insert(tx, priority).await?;
-        }
-        Ok(())
+    pub(crate) async fn insert_all(&self, txs: Vec<(EnqueuedTransaction, TransactionPriority)>) {
+        self.inner.write().await.extend(txs);
     }
 
     /// pops the transaction with the highest priority from the mempool
     #[must_use]
-    pub(crate) async fn pop(&mut self) -> Option<(SignedTransaction, TransactionPriority)> {
-        let mut inner = self.inner.lock().await;
-        let (hash, priority) = inner.queue.pop_max()?;
-        let tx = inner.hash_to_tx.remove(&hash)?;
-        Some((tx, priority))
+    pub(crate) async fn pop(&self) -> Option<(EnqueuedTransaction, TransactionPriority)> {
+        self.inner.write().await.pop()
     }
 
     /// removes a transaction from the mempool
-    pub(crate) async fn remove(&mut self, tx_hash: &[u8; 32]) {
-        let mut inner = self.inner.lock().await;
-        inner.queue.remove(tx_hash);
-        inner.hash_to_tx.remove(tx_hash);
+    pub(crate) async fn remove(&self, tx_hash: [u8; 32]) {
+        let enqueued_tx = EnqueuedTransaction {
+            tx_hash,
+            signed_tx: dummy_signed_tx().clone(),
+        };
+        self.inner.write().await.remove(&enqueued_tx);
     }
 
-    /// removes all the given transactions from the mempool
-    pub(crate) async fn remove_all(&mut self, tx_hashes: &[[u8; 32]]) {
-        for tx_hash in tx_hashes {
-            self.remove(tx_hash).await;
+    /// Updates the priority of the txs in the mempool based on the current state, and removes any
+    /// that are now invalid.
+    ///
+    /// NOTE: this function locks the mempool until every tx has been checked. This could
+    /// potentially stall consensus from moving to the next round if the mempool is large.
+    pub(crate) async fn update_priorities<F, O>(
+        &self,
+        current_account_nonce_getter: F,
+    ) -> anyhow::Result<()>
+    where
+        F: Fn(Address) -> O,
+        O: Future<Output = anyhow::Result<u32>>,
+    {
+        let mut txs_to_remove = Vec::new();
+        let mut current_account_nonces = HashMap::new();
+
+        let mut queue = self.inner.write().await;
+        for (enqueued_tx, priority) in queue.iter_mut() {
+            let address = enqueued_tx.address();
+            // Try to get the current account nonce from the ones already retrieved.
+            let current_account_nonce = if let Some(nonce) = current_account_nonces.get(&address) {
+                *nonce
+            } else {
+                // Fall back to getting via the getter and adding it to the local temp collection.
+                let nonce = current_account_nonce_getter(enqueued_tx.address())
+                    .await
+                    .context("failed to fetch account nonce")?;
```

Comment:
Good point - seems like it might be fine to continue, but I wanted to match [the existing behaviour in `main`](https://github.com/astriaorg/astria/blob/691888bc5c3daf4dcbb243734f11b88d48569a7e/crates/astria-sequencer/src/app/mod.rs#L1149).

@noot - do you have an opinion on this?

Reply:
not being able to fetch the current nonce would mean that the db was unreadable or what's stored at that value is unreadable, both of which are pretty critical errors. i think returning early makes sense here